tests/compile-fail/non_extern_foreign_call_abi_mismatch.rs

```diff
@@ -0,0 +1,22 @@
+// revisions: no_cache cache
+
+#[no_mangle]
+fn foo() {}
+
+fn main() {
+    #[cfg(cache)]
+    {
+        // `Instance` caching should not suppress ABI check.
+        extern "Rust" {
+            fn foo();
+        }
+        unsafe { foo() }
+    }
+    #[cfg_attr(cache, allow(clashing_extern_declarations))]
+    extern "C" {
+        fn foo();
+    }
+    unsafe { foo() }
+    //[no_cache]~^ ERROR Undefined Behavior: calling a function with ABI Rust using caller ABI C
+    //[cache]~^^ ERROR Undefined Behavior: calling a function with ABI Rust using caller ABI C
+}
```

Comment:
These tests here look a lot like the tests we have in `function_pointers` so far... maybe rename that folder to `function_calls` and move (most of) the new tests there? The `check_arg` tests would also belong here.